pycloudlib/qemu/cloud.py

```diff
@@ -0,0 +1,634 @@
+            "UTF-8",
+            str(user_data_path),
+            str(meta_data_path),
+        ]
+
+        # Vendor data is entirely optional, so don't create it automatically
+        if vendor_data:
+            vendor_data_path = instance_dir / "vendor-data"
+            vendor_data_path.write_text(vendor_data, encoding="utf-8")
+            args.append(str(vendor_data_path))
+
+        subprocess.run(
+            args,
+            check=True,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+        )
+        return iso_path
+
+    def _create_qcow_image(self, instance_path: Path, base_image: Path):
+        """Create a qcow image from original iso.
+
+        This is so we can make cheap destructive changes without modifying
+        the original image. Also add a GB to the image size so we can
+        we have enough space to write changes.
+
+        Args:
+            instance_path: Path, path to instance image
+            base_image: Path, path to base image
+        """
+        subprocess.run(
+            [
+                "qemu-img",
+                "create",
+                "-f",
+                "qcow2",
+                "-b",
+                base_image,
+                "-F",
+                "qcow2",
+                str(instance_path),
+                "10G",  # This space is only taken when needed
+            ],
+            check=True,
+        )
+
+    def _get_ubuntu_kernel_from_image_dir(self, image: Path) -> Optional[Path]:
+        kernels = list(
+            image.parent.glob("*-server-cloudimg-amd64-vmlinuz-generic")
+        )
+        if len(kernels) != 1:
+            self._log.warning("Unable to find kernel for image: %s", image)
+            return None
+        return kernels[0].absolute()
+
+    def _find_base_image(self, image_id: str) -> Path:
+        if Path(image_id).exists():
+            base_image = Path(image_id)
+        elif Path(self.image_dir, image_id).exists():
+            base_image = Path(self.image_dir, image_id)
+        else:
+            raise ImageNotFoundError(
+                resource_id=image_id,
+                message=(
+                    f"Could not find '{image_id}' as absolute path "
+                    f"or in '{self.image_dir}'."
+                ),
+            )
+        self._log.debug("Using base image: %s", base_image)
+        return base_image
+
+    def _parse_instance_type(self, instance_type: str) -> Tuple[int, int]:
+        cpu_mem = re.search(r"^c(?P<cpus>\d+)m(?P<memory>\d+$)", instance_type)
+        if not cpu_mem:
+            raise ValueError(
+                "instance_type must be in the form of c#m#, not "
+                f"{instance_type}"
+            )
+        return int(cpu_mem["cpus"]), int(cpu_mem["memory"])
+
+    def _get_kernel_path(
+        self, kernel_path, kernel_cmdline: str, image_id, base_image
+    ) -> Optional[Path]:
+        kernel = None
+        if kernel_cmdline and not kernel_path:
```

Comment:
We are already performing this check outside this scope at [the call-site](https://github.com/canonical/pycloudlib/pull/323/files#diff-9c031a07867513f1651aa8c522f0ca475eed8b5756bafa4ebaa92966811fdb57R511). Do you want to leave this check in here and remove the one at the call-site and we can just set or return the original `kernel_path` from this method?